src/decoder/transform/palette.rs

```diff
@@ -0,0 +1,361 @@
+//! Helpers for taking a slice of indeces (indices into `PLTE` and/or `trNS`
+//! entries) and transforming this into RGB or RGBA output.
+//!
+//! # Memoization
+//!
+//! To achieve higher throughput, `create_rgba_palette` combines entries from
+//! `PLTE` and `trNS` chunks into a single lookup table.  This is based on the
+//! ideas explored in https://crbug.com/706134.
+//!
+//! Memoization is a trade-off:
+//! * On one hand, memoization requires spending X ns before starting to call
+//!   `expand_paletted_...` functions.
+//! * On the other hand, memoization improves the throughput of the
+//!   `expand_paletted_...` functions - they take Y ns less to process each byte
+//!
+//! Based on X and Y, we can try to calculate the breakeven point.  It seems
+//! that memoization is a net benefit for images bigger than around 13x13 pixels.
+
+use super::{unpack_bits, TransformFn};
+use crate::{BitDepth, Info};
+
+pub fn create_expansion_into_rgb8(info: &Info) -> TransformFn {
+    let rgba_palette = create_rgba_palette(info);
+
+    if info.bit_depth == BitDepth::Eight {
+        Box::new(move |input, output, _info| expand_8bit_into_rgb8(input, output, &rgba_palette))
+    } else {
+        Box::new(move |input, output, info| expand_into_rgb8(input, output, info, &rgba_palette))
+    }
+}
+
+pub fn create_expansion_into_rgba8(info: &Info) -> TransformFn {
+    let rgba_palette = create_rgba_palette(info);
+    Box::new(move |input, output, info| {
+        expand_paletted_into_rgba8(input, output, info, &rgba_palette)
+    })
+}
+
+fn create_rgba_palette(info: &Info) -> [[u8; 4]; 256] {
+    let palette = info.palette.as_deref().expect("Caller should verify");
+    let trns = info.trns.as_deref().unwrap_or(&[]);
+
+    // > The tRNS chunk shall not contain more alpha values than there are palette
+    // entries, but a tRNS chunk may contain fewer values than there are palette
+    // entries. In this case, the alpha value for all remaining palette entries is
+    // assumed to be 255.
+    //
+    // It seems, accepted reading is to fully *ignore* an invalid tRNS as if it were
+    // completely empty / all pixels are non-transparent.
+    let trns = if trns.len() <= palette.len() / 3 {
+        trns
+    } else {
+        &[]
+    };
+
+    // Default to black, opaque entries.
+    let mut rgba_palette = [[0, 0, 0, 0xFF]; 256];
+
+    // Copy `palette` (RGB) entries into `rgba_palette`.  This may clobber alpha
+    // values in `rgba_palette` - we need to fix this later.
+    {
+        let mut palette_iter = palette;
+        let mut rgba_iter = &mut rgba_palette[..];
+        while palette_iter.len() >= 4 {
```

Comment:
```suggestion
        while palette_iter.len() >= 4 && !rgba_iter.is_empty() {
```

Modifying https://godbolt.org/z/KeKrfWcs5, I haven't benchmarked or tested it however.
Removes the panic but doesn't unroll the loop like the original https://rust.godbolt.org/z/53MaeEano

Reply:
This results in a 25% regression.

The compiler is aware that the branches leading to the panic are cold, and so decides to unroll the loop. The CPU should speculate right past the branch leading to a panic. So it is best to keep it as a clearly-cold panic rather than a regular branch.